app/usecase/event/dispatcher.go

```diff
@@ -0,0 +1,79 @@
+package event
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/asaskevich/EventBus"
+)
+
+var _ Dispatcher = (*EventDispatcher)(nil)
+var _ Subscriber = (*EventDispatcher)(nil)
+
+// ErrDispatcherIsClosed tells that there is no way to perform manipulations with event dispatcher
+var ErrDispatcherIsClosed = errors.New("failed to perform the operation, the dispatcher is closed")
+
+// EventDispatcher publishes an event to all its subscribers
+type EventDispatcher struct {
+	bus      EventBus.Bus
+	lock     sync.RWMutex
+	isClosed bool
+}
+
+func (d *EventDispatcher) Dispatch(event Event) error {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+
+	if d.isClosed {
+		return ErrDispatcherIsClosed
+	}
+
+	d.bus.Publish(event.GetName(), event)
+
+	return nil
+}
+
+func (d *EventDispatcher) Subscribe(eventName string, listener Listener) error {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+
+	if d.isClosed {
+		return ErrDispatcherIsClosed
+	}
+
+	return d.bus.SubscribeAsync(eventName, listener.Handle, false)
```

Comment:
It seems that the third parameter has the concept of sequential execution for transactions. This may be inspirational for us to build the transaction infrastructure.

Reply:
I wonder what's the difference between `Subscribe`  and `SubscribeAsync` here. Is `Subscribe` waiting the call back to finish executing before it starts processing the second event?

Reply:
`SubscribeAsync` informs corresponding listeners about the incoming event in each separate `goroutine`

`Subscribe` uses the opposite strategy